src/cpp/src/whisper/audio_processing.cpp

```diff
@@ -0,0 +1,343 @@
+            fft_in[j] = hann[j] * samples[offset + j];
+        }
+        // fill the rest with zeros
+        if (n_samples - offset < frame_size) {
+            std::fill(fft_in.begin() + (n_samples - offset), fft_in.end(), 0.0);
+        }
+
+        // FFT
+        fft(fft_in, fft_out);
+
+        // Calculate modulus^2 of complex numbers
+        // Use pow(fft_out[2 * j + 0], 2) + pow(fft_out[2 * j + 1], 2) causes inference quality problem? Interesting.
+        for (int j = 0; j < n_fft; j++) {
+            fft_out[j] = (fft_out[2 * j + 0] * fft_out[2 * j + 0] + fft_out[2 * j + 1] * fft_out[2 * j + 1]);
+        }
+
+        // mel spectrogram
+        for (int j = 0; j < mel.n_mel; j++) {
+            double sum = 0.0;
+
+            // unroll loop (suggested by GH user @lunixbochs)
+            int k = 0;
+            for (k = 0; k < n_fft - 3; k += 4) {
+                sum += fft_out[k + 0] * filters.data[j * n_fft + k + 0] +
+                       fft_out[k + 1] * filters.data[j * n_fft + k + 1] +
+                       fft_out[k + 2] * filters.data[j * n_fft + k + 2] +
+                       fft_out[k + 3] * filters.data[j * n_fft + k + 3];
+            }
+
+            // handle n_fft remainder
+            for (; k < n_fft; k++) {
+                sum += fft_out[k] * filters.data[j * n_fft + k];
+            }
+
+            sum = log10(std::max(sum, 1e-10));
+
+            mel.data[j * mel.n_len + i] = sum;
+        }
+    }
+
+    // Otherwise fft_out are all zero
+    double sum = log10(1e-10);
+    for (; i < mel.n_len; i += n_threads) {
+        for (int j = 0; j < mel.n_mel; j++) {
+            mel.data[j * mel.n_len + i] = sum;
+        }
+    }
+}
+}  // namespace
+
+namespace ov {
+namespace genai {
+namespace utils {
+namespace audio {
+
+// In FFT, we frequently use sine and cosine operations with the same values.
+// We can use precalculated values to speed up the process.
+void fill_sin_cos_table() {
+    static bool is_filled = false;
+    if (is_filled)
+        return;
+    for (int i = 0; i < SIN_COS_N_COUNT; i++) {
+        double theta = (2 * M_PI * i) / SIN_COS_N_COUNT;
+        sin_vals[i] = sinf(theta);
+        cos_vals[i] = cosf(theta);
+    }
+    is_filled = true;
+}
+
+std::vector<float> mel_spectrogram_convert_audio(const std::vector<float> pcmf32,
+                                                 const int sample_rate,
+                                                 const int frame_size,
+                                                 const int frame_step,
+                                                 const int n_threads) {
+    const float* samples = pcmf32.data();
+    const int n_samples = pcmf32.size();
+    whisper_filters filters;
+    whisper_mel mel;
+
+    std::ifstream file_fdata;
+    // todo: handle path. Copy into mel_filters_data.bin into build, install dirs
+    file_fdata.open("./assets/whisper/mel_filters_data.bin", std::ios::binary);
```

Comment:
is this file going to be a part of genAI package?
If no, we cannot read it here w/ relative path

Reply:
I plan to remove this file and compute `mel_filters_data` on initialization